fastddsspy_participants/src/cpp/visualization/ModelParser.cpp

```diff
@@ -211,68 +211,282 @@ ComplexParticipantData ModelParser::participants(const SpyModel& model, const dd
+
+    for (const auto& endpoint : model.endpoint_database_)
+    {
+        if (endpoint.first == guid && endpoint.second.is_reader())
+        {
+            // Look for participant name
+            for (const auto& participant : model.participant_database_)
+            {
+                if (participant.first.guid_prefix() == guid.guid_prefix())
+                {
+                    result.participant_name = participant.second.name;
+                    break;
+                }
+            }
+            set_complex_endpoint(result, endpoint.second);
+            break;
+        }
+    }
+
     return result;
 }
 
+/*
+ * This is an auxiliary function that is used to get topics endpoint database
+ */
+std::vector<eprosima::ddspipe::core::types::DdsTopic> get_topics(const SpyModel& model)
+{
+    std::vector<eprosima::ddspipe::core::types::DdsTopic> result;
+    for (const auto& endpoint : model.endpoint_database_)
+    {
+        if (std::find(result.begin(), result.end(), endpoint.second.topic) == result.end())
+        {
+            result.push_back(endpoint.second.topic);
+        }
+    }
+    return result;
+}
 
 std::vector<SimpleTopicData> ModelParser::topics(const SpyModel& model)
 {
     std::vector<SimpleTopicData> result;
-    // TODO
+
+    std::vector<eprosima::ddspipe::core::types::DdsTopic> endpoints_topics = get_topics(model);
+    for (const auto& topic : endpoints_topics)
+    {
+        int datawriters = 0;
+        int datareaders = 0;
+        for (const auto& endpoint : model.endpoint_database_)
+        {
+            if (endpoint.second.topic.m_topic_name == topic.m_topic_name)
+            {
+                if (endpoint.second.is_reader())
+                {
+                    datareaders++;
+                }
+                if (endpoint.second.is_writer())
+                {
+                    datawriters++;
+                }
+            }
+        }
+        result.push_back({
+                    topic.m_topic_name,
+                    topic.type_name,
+                    datawriters,
+                    datareaders,
+                    {
+                        10,     // TODO not implementated yet
+                        "Hz"
+                    }
+                });
+    }
+
     return result;
 }
 
 std::vector<ComplexTopicData> ModelParser::topics_verbose(const SpyModel& model)
 {
     std::vector<ComplexTopicData> result;
-    // TODO
+
+    std::vector<eprosima::ddspipe::core::types::DdsTopic> endpoints_topics = get_topics(model);
+    for (const auto& topic : endpoints_topics)
+    {
+        result.push_back(topics(model, topic.m_topic_name));
+    }
+
     return result;
 }
 
 ComplexTopicData ModelParser::topics(const SpyModel& model, const std::string& topic_name)
 {
     ComplexTopicData result;
-    // TODO
+    std::vector<ComplexTopicData::Endpoint> datareaders;
+    std::vector<ComplexTopicData::Endpoint> datawriters;
+
+    std::string type_name = "Not discovered yet";
```

Comment:
NIT: Maybe an easier way is to use function get_topic() and get wether this topic exist or not. If not return, if it does, then fill the fields and do the writer and reader search.